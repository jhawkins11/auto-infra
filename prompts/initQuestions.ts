import { InitAnswers, PromptQuestion } from '../types/prompts.js';

export const initQuestions: PromptQuestion<InitAnswers>[] = [
  {
    type: 'input',
    name: 'projectName',
    message: 'Project name:',
    default: 'my-project',
  },
  {
    type: 'input',
    name: 'awsRegion',
    message: 'AWS region:',
    default: 'us-east-1',
  },
  {
    type: 'input',
    name: 'domain',
    message: 'Domain (optional):',
    default: '',
  },
  {
    type: 'list',
    name: 'dbSize',
    message: 'Database size:',
    choices: ['small', 'medium', 'large'],
    default: 'small',
  },
];
