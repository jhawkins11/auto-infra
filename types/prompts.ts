export type InitAnswers = {
  projectName: string;
  awsRegion: string;
  domain?: string;
  containerImageUrl?: string;
  containerPort?: number;
  ecsTaskSize?: 'small' | 'medium' | 'large';
  desiredCount?: number;
  healthCheckPath?: string;
  dbEngine?: 'postgresql' | 'mysql';
  dbInstanceSize?: 'micro' | 'small' | 'medium' | 'large' | 'xlarge';
  dbUsername?: string;
  dbPassword?: string;
};

export type PromptQuestion<T> = {
  type: string;
  name: keyof T | string;
  message: string;
  default?: string | ((answers: T) => string);
  choices?: string[];
  validate?: (input: string) => boolean | string;
  filter?: (input: string) => string;
  mask?: string | boolean;
};
