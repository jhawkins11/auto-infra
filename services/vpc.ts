import { InitAnswers } from '../types/prompts.js';
import {
  ensureDirectoryExists,
  ensureGitignoreHasTerraformIgnores,
  writeConfigFile,
} from './fileWriter.js';

export const generateVpcConfig = async (
  answers: InitAnswers,
): Promise<void> => {
  const configDirectory = `./${answers.projectName}`;
  const vpcConfigPath = `${configDirectory}/vpc.tfvars`;

  const vpcConfig = {
    project_name: answers.projectName,
    aws_region: answers.awsRegion,
    ...(answers.domain && { domain_name: answers.domain }),
  };

  await ensureDirectoryExists(vpcConfigPath);
  await writeConfigFile(vpcConfigPath, vpcConfig);
  await ensureGitignoreHasTerraformIgnores(configDirectory);
  console.info('VPC configuration generated.');
};
