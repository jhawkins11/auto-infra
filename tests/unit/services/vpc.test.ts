jest.mock('node:fs/promises', () => ({
  writeFile: jest.fn(),
  mkdir: jest.fn(),
  readFile: jest.fn().mockRejectedValue(new Error('not found')),
}));

import * as fs from 'node:fs/promises';

import { generateVpcConfig } from '../../../services/vpc.js';
import { InitAnswers } from '../../../types/prompts.js';

describe('generateVpcConfig', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should generate a vpc.tfvars file with the correct content', async () => {
    const answers = {
      projectName: 'test-project',
      awsRegion: 'us-west-2',
      domain: 'example.org',
    } as InitAnswers;

    await generateVpcConfig(answers);

    const mkdirMock = fs.mkdir as unknown as jest.Mock;
    const writeFileMock = fs.writeFile as unknown as jest.Mock;

    expect(mkdirMock).toHaveBeenCalledTimes(1);
    expect(mkdirMock).toHaveBeenCalledWith(`./${answers.projectName}`, {
      recursive: true,
    });

    // find the writeFile call that wrote the vpc.tfvars file
    const vpcCall = writeFileMock.mock.calls.find((c: any[]) =>
      String(c[0]).endsWith('/vpc.tfvars'),
    );
    expect(vpcCall).toBeDefined();

    const [pathArg, contentArg] = vpcCall;
    expect(pathArg).toBe(`./${answers.projectName}/vpc.tfvars`);
    expect(contentArg).toMatchSnapshot();
  });
});
