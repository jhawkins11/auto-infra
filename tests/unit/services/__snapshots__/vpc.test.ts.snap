// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generateVpcConfig should generate a vpc.tfvars file with the correct content 1`] = `
"project_name = "test-project"
aws_region = "us-west-2"
domain_name = "example.org""
`;
